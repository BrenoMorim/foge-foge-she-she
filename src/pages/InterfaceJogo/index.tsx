import { useParams } from "react-router-dom";
import MenuPontuacao from "components/MenuPontuacao";
import Mapa from "components/Mapa";
import MenuComandos from "components/MenuComandos";
import { lazy, useState } from "react";
import IFase from "types/IFase";
import ManipuladorDeState from "util/ManipuladorDeState";
import TelaFinal from "components/TelaFinal";
import carregaFasePorId from "util/carregaFasePorId";

const NotFound = lazy(() => import('pages/NotFound'));

export default function InterfaceJogo() {
  const params = useParams();
  const id = params.id ? parseInt(params.id) : 1;
  const faseEncontrada = carregaFasePorId(id);
  if (faseEncontrada === undefined) return <NotFound />;
  const [fase, setFase] = useState<IFase>(faseEncontrada);

  const jogoAcabou = (fase.ganhou || fase.perdeu);
  const manipuladorDeState = new ManipuladorDeState(fase, setFase);

  return (
    <section>
      {(jogoAcabou) && <TelaFinal manipuladorDeState={manipuladorDeState} pontuacaoFinal={fase.pontuacao} vitoria={fase.ganhou}/>}
      <MenuPontuacao labelFase={fase.label} pontuacao={fase.pontuacao}/>
      <Mapa mapa={fase.mapa}/>
      <MenuComandos manipuladorDeState={manipuladorDeState} faseAtual={fase}/>
    </section>
  );
}
