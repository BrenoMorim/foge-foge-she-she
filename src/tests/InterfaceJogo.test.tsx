import { render, screen } from "@testing-library/react";
import InterfaceJogo from "pages/InterfaceJogo";
import React from "react";

const mockParams = jest.fn().mockReturnValue({id: 1});
jest.mock('react-router-dom', () => {
    return {
        useParams: () => mockParams
    }
});

describe('Na interface do jogo', () => {

    test('Deve ser renderizada corretamente', () => {
        const interfaceJogo = render(<InterfaceJogo/>);
        expect(interfaceJogo).toMatchSnapshot();
    });

    describe('Ao ser carregada pela primeira vez', () => {

        beforeEach(() => render(<InterfaceJogo/>));

        test('Deve conter as informações corretas', () => {
            const faseLabel = screen.getByTestId('fase-label');
            expect(faseLabel).toHaveTextContent('Tutorial');

            const pontuacao = screen.getByTestId('fase-pontuacao');
            expect(pontuacao).toHaveTextContent("0");
        });

        test('Deve conter personagens e outros elementos no mapa', () => {
            const sherlock = screen.getByAltText('sherlock');
            expect(sherlock).toBeInTheDocument();

            const breno = screen.getByAltText('breno');
            expect(breno).toBeInTheDocument();

            const jaminha = screen.getByAltText('jaminha');
            expect(jaminha).toBeInTheDocument();

            const especial = screen.getByAltText('especial');
            expect(especial).toBeInTheDocument();

            const racoes = screen.getAllByAltText('racao');
            expect(racoes).toHaveLength(4);
        });

        test('Deve conter botões de movimento, desabilitados quando necessário', () => {
            const botoes = screen.getAllByRole('button');
            expect(botoes).toHaveLength(5);

            const botaoEspecial = screen.getByTestId('especial');
            expect(botaoEspecial).toBeDisabled();

            const botaoCima = screen.getByTestId('cima');
            expect(botaoCima).toBeDisabled();
            const botaoBaixo = screen.getByTestId('baixo');
            expect(botaoBaixo).toBeEnabled();
        });
    });
});