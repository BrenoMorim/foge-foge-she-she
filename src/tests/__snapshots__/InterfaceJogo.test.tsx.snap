// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Na interface do jogo Deve ser renderizada corretamente 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section>
        <section
          class="pontuacao base__painel-roxo"
        >
          <article
            class="base__flex-container"
          >
            <h2
              class="pontuacao__subtitulo"
            >
              Fase:
            </h2>
            <h2
              class="pontuacao__subtitulo"
              data-testid="fase-label"
            >
              Tutorial
            </h2>
          </article>
          <article
            class="base__flex-container"
          >
            <h2
              class="pontuacao__subtitulo"
            >
              Pontuação:
            </h2>
            <h2
              class="pontuacao__subtitulo"
              data-testid="fase-pontuacao"
            >
              0
            </h2>
          </article>
        </section>
        <table
          class="mapa__container"
        >
          <tbody
            class="mapa"
          >
            <tr
              class="mapa__linha"
            >
              <td
                class="mapa__pixel"
              >
                <img
                  alt="espaco-vazio"
                  class="pixel__espaco-vazio"
                  src="/assets/mapa/espaco-vazio.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="espaco-vazio"
                  class="pixel__espaco-vazio"
                  src="/assets/mapa/espaco-vazio.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="espaco-vazio"
                  class="pixel__espaco-vazio"
                  src="/assets/mapa/espaco-vazio.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="canto"
                  class="pixel__canto-superior-esquerdo"
                  src="/assets/mapa/canto.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-horizontal"
                  src="/assets/mapa/parede.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="canto"
                  class="pixel__canto-superior-direito"
                  src="/assets/mapa/canto.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="espaco-vazio"
                  class="pixel__espaco-vazio"
                  src="/assets/mapa/espaco-vazio.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="espaco-vazio"
                  class="pixel__espaco-vazio"
                  src="/assets/mapa/espaco-vazio.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="espaco-vazio"
                  class="pixel__espaco-vazio"
                  src="/assets/mapa/espaco-vazio.svg"
                />
              </td>
            </tr>
            <tr
              class="mapa__linha"
            >
              <td
                class="mapa__pixel"
              >
                <img
                  alt="canto"
                  class="pixel__canto-superior-esquerdo"
                  src="/assets/mapa/canto.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-horizontal"
                  src="/assets/mapa/parede.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-horizontal"
                  src="/assets/mapa/parede.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="canto"
                  class="pixel__canto-inferior-direito"
                  src="/assets/mapa/canto.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="sherlock"
                  class="pixel__personagem"
                  src="/assets/mapa/sherlock.png"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="canto"
                  class="pixel__canto-inferior-esquerdo"
                  src="/assets/mapa/canto.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-horizontal"
                  src="/assets/mapa/parede.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-horizontal"
                  src="/assets/mapa/parede.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="canto"
                  class="pixel__canto-superior-direito"
                  src="/assets/mapa/canto.svg"
                />
              </td>
            </tr>
            <tr
              class="mapa__linha"
            >
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-vertical"
                  src="/assets/mapa/parede.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="jaminha"
                  class="pixel__personagem"
                  src="/assets/mapa/jaminha.png"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="racao"
                  class="pixel__racao"
                  src="/assets/mapa/racao.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="racao"
                  class="pixel__racao"
                  src="/assets/mapa/racao.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="especial"
                  class="pixel__especial"
                  src="/assets/mapa/especial.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="racao"
                  class="pixel__racao"
                  src="/assets/mapa/racao.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="racao"
                  class="pixel__racao"
                  src="/assets/mapa/racao.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="breno"
                  class="pixel__personagem"
                  src="/assets/mapa/breno.png"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-vertical"
                  src="/assets/mapa/parede.svg"
                />
              </td>
            </tr>
            <tr
              class="mapa__linha"
            >
              <td
                class="mapa__pixel"
              >
                <img
                  alt="canto"
                  class="pixel__canto-inferior-esquerdo"
                  src="/assets/mapa/canto.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-horizontal"
                  src="/assets/mapa/parede.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-horizontal"
                  src="/assets/mapa/parede.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-horizontal"
                  src="/assets/mapa/parede.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-horizontal"
                  src="/assets/mapa/parede.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-horizontal"
                  src="/assets/mapa/parede.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-horizontal"
                  src="/assets/mapa/parede.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="parede"
                  class="pixel__parede-horizontal"
                  src="/assets/mapa/parede.svg"
                />
              </td>
              <td
                class="mapa__pixel"
              >
                <img
                  alt="canto"
                  class="pixel__canto-inferior-direito"
                  src="/assets/mapa/canto.svg"
                />
              </td>
            </tr>
          </tbody>
        </table>
        <section
          class="comandos"
        >
          <button
            class="comandos__botao"
            data-testid="cima"
            disabled=""
          >
            <img
              alt="cima"
              class="botao__imagem botao__imagem--cima"
              src="/assets/botoes/seta.svg"
            />
          </button>
          <button
            class="comandos__botao"
            data-testid="baixo"
          >
            <img
              alt="baixo"
              class="botao__imagem botao__imagem--baixo"
              src="/assets/botoes/seta.svg"
            />
          </button>
          <button
            class="comandos__botao"
            data-testid="esquerda"
            disabled=""
          >
            <img
              alt="esquerda"
              class="botao__imagem botao__imagem--esquerda"
              src="/assets/botoes/seta.svg"
            />
          </button>
          <button
            class="comandos__botao"
            data-testid="direita"
            disabled=""
          >
            <img
              alt="direita"
              class="botao__imagem botao__imagem--direita"
              src="/assets/botoes/seta.svg"
            />
          </button>
          <button
            class="comandos__botao"
            data-testid="especial"
            disabled=""
          >
            <img
              alt="Especial"
              class="botao__imagem"
              src="/assets/botoes/especial.svg"
            />
          </button>
        </section>
      </section>
    </div>
  </body>,
  "container": <div>
    <section>
      <section
        class="pontuacao base__painel-roxo"
      >
        <article
          class="base__flex-container"
        >
          <h2
            class="pontuacao__subtitulo"
          >
            Fase:
          </h2>
          <h2
            class="pontuacao__subtitulo"
            data-testid="fase-label"
          >
            Tutorial
          </h2>
        </article>
        <article
          class="base__flex-container"
        >
          <h2
            class="pontuacao__subtitulo"
          >
            Pontuação:
          </h2>
          <h2
            class="pontuacao__subtitulo"
            data-testid="fase-pontuacao"
          >
            0
          </h2>
        </article>
      </section>
      <table
        class="mapa__container"
      >
        <tbody
          class="mapa"
        >
          <tr
            class="mapa__linha"
          >
            <td
              class="mapa__pixel"
            >
              <img
                alt="espaco-vazio"
                class="pixel__espaco-vazio"
                src="/assets/mapa/espaco-vazio.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="espaco-vazio"
                class="pixel__espaco-vazio"
                src="/assets/mapa/espaco-vazio.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="espaco-vazio"
                class="pixel__espaco-vazio"
                src="/assets/mapa/espaco-vazio.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="canto"
                class="pixel__canto-superior-esquerdo"
                src="/assets/mapa/canto.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-horizontal"
                src="/assets/mapa/parede.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="canto"
                class="pixel__canto-superior-direito"
                src="/assets/mapa/canto.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="espaco-vazio"
                class="pixel__espaco-vazio"
                src="/assets/mapa/espaco-vazio.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="espaco-vazio"
                class="pixel__espaco-vazio"
                src="/assets/mapa/espaco-vazio.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="espaco-vazio"
                class="pixel__espaco-vazio"
                src="/assets/mapa/espaco-vazio.svg"
              />
            </td>
          </tr>
          <tr
            class="mapa__linha"
          >
            <td
              class="mapa__pixel"
            >
              <img
                alt="canto"
                class="pixel__canto-superior-esquerdo"
                src="/assets/mapa/canto.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-horizontal"
                src="/assets/mapa/parede.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-horizontal"
                src="/assets/mapa/parede.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="canto"
                class="pixel__canto-inferior-direito"
                src="/assets/mapa/canto.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="sherlock"
                class="pixel__personagem"
                src="/assets/mapa/sherlock.png"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="canto"
                class="pixel__canto-inferior-esquerdo"
                src="/assets/mapa/canto.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-horizontal"
                src="/assets/mapa/parede.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-horizontal"
                src="/assets/mapa/parede.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="canto"
                class="pixel__canto-superior-direito"
                src="/assets/mapa/canto.svg"
              />
            </td>
          </tr>
          <tr
            class="mapa__linha"
          >
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-vertical"
                src="/assets/mapa/parede.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="jaminha"
                class="pixel__personagem"
                src="/assets/mapa/jaminha.png"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="racao"
                class="pixel__racao"
                src="/assets/mapa/racao.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="racao"
                class="pixel__racao"
                src="/assets/mapa/racao.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="especial"
                class="pixel__especial"
                src="/assets/mapa/especial.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="racao"
                class="pixel__racao"
                src="/assets/mapa/racao.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="racao"
                class="pixel__racao"
                src="/assets/mapa/racao.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="breno"
                class="pixel__personagem"
                src="/assets/mapa/breno.png"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-vertical"
                src="/assets/mapa/parede.svg"
              />
            </td>
          </tr>
          <tr
            class="mapa__linha"
          >
            <td
              class="mapa__pixel"
            >
              <img
                alt="canto"
                class="pixel__canto-inferior-esquerdo"
                src="/assets/mapa/canto.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-horizontal"
                src="/assets/mapa/parede.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-horizontal"
                src="/assets/mapa/parede.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-horizontal"
                src="/assets/mapa/parede.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-horizontal"
                src="/assets/mapa/parede.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-horizontal"
                src="/assets/mapa/parede.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-horizontal"
                src="/assets/mapa/parede.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="parede"
                class="pixel__parede-horizontal"
                src="/assets/mapa/parede.svg"
              />
            </td>
            <td
              class="mapa__pixel"
            >
              <img
                alt="canto"
                class="pixel__canto-inferior-direito"
                src="/assets/mapa/canto.svg"
              />
            </td>
          </tr>
        </tbody>
      </table>
      <section
        class="comandos"
      >
        <button
          class="comandos__botao"
          data-testid="cima"
          disabled=""
        >
          <img
            alt="cima"
            class="botao__imagem botao__imagem--cima"
            src="/assets/botoes/seta.svg"
          />
        </button>
        <button
          class="comandos__botao"
          data-testid="baixo"
        >
          <img
            alt="baixo"
            class="botao__imagem botao__imagem--baixo"
            src="/assets/botoes/seta.svg"
          />
        </button>
        <button
          class="comandos__botao"
          data-testid="esquerda"
          disabled=""
        >
          <img
            alt="esquerda"
            class="botao__imagem botao__imagem--esquerda"
            src="/assets/botoes/seta.svg"
          />
        </button>
        <button
          class="comandos__botao"
          data-testid="direita"
          disabled=""
        >
          <img
            alt="direita"
            class="botao__imagem botao__imagem--direita"
            src="/assets/botoes/seta.svg"
          />
        </button>
        <button
          class="comandos__botao"
          data-testid="especial"
          disabled=""
        >
          <img
            alt="Especial"
            class="botao__imagem"
            src="/assets/botoes/especial.svg"
          />
        </button>
      </section>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
